// @celo/base/lib/currencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.currencyToShortMap = exports.SHORT_CURRENCIES = exports.resolveCurrency = exports.CURRENCIES = exports.Token = exports.StableToken = exports.CURRENCY_ENUM = void 0;
/** @deprecated use StableToken and Token */
var CURRENCY_ENUM;
(function (CURRENCY_ENUM) {
    CURRENCY_ENUM["GOLD"] = "Celo Gold";
    CURRENCY_ENUM["DOLLAR"] = "Celo Dollar";
    CURRENCY_ENUM["EURO"] = "Celo Euro";
})(CURRENCY_ENUM || (exports.CURRENCY_ENUM = CURRENCY_ENUM = {}));
var StableToken;
(function (StableToken) {
    StableToken["cUSD"] = "cUSD";
    StableToken["cEUR"] = "cEUR";
    StableToken["cREAL"] = "cREAL";
})(StableToken || (exports.StableToken = StableToken = {}));
var Token;
(function (Token) {
    Token["CELO"] = "CELO";
})(Token || (exports.Token = Token = {}));
/** @deprecated */
exports.CURRENCIES = {
    [CURRENCY_ENUM.GOLD]: {
        symbol: '',
        code: 'cGLD',
        displayDecimals: 3,
    },
    [CURRENCY_ENUM.DOLLAR]: {
        symbol: '$',
        code: 'cUSD',
        displayDecimals: 2,
    },
    [CURRENCY_ENUM.EURO]: {
        symbol: '€',
        code: 'cEUR',
        displayDecimals: 2,
    },
};
const resolveCurrency = (label) => {
    if (label && label.toLowerCase().includes('dollar')) {
        return CURRENCY_ENUM.DOLLAR;
    }
    else if (label && label.toLowerCase().includes('euro')) {
        return CURRENCY_ENUM.EURO;
    }
    else if (label && label.toLowerCase().includes('gold')) {
        return CURRENCY_ENUM.GOLD;
    }
    else {
        console.info('Unable to resolve currency from label: ' + label);
        return CURRENCY_ENUM.DOLLAR;
    }
};
exports.resolveCurrency = resolveCurrency;
/** @deprecated use StableToken and Token */
var SHORT_CURRENCIES;
(function (SHORT_CURRENCIES) {
    SHORT_CURRENCIES["DOLLAR"] = "dollar";
    SHORT_CURRENCIES["GOLD"] = "gold";
    SHORT_CURRENCIES["EURO"] = "euro";
})(SHORT_CURRENCIES || (exports.SHORT_CURRENCIES = SHORT_CURRENCIES = {}));
/** @deprecated use StableToken and Token */
exports.currencyToShortMap = {
    [CURRENCY_ENUM.DOLLAR]: SHORT_CURRENCIES.DOLLAR,
    [CURRENCY_ENUM.GOLD]: SHORT_CURRENCIES.GOLD,
    [CURRENCY_ENUM.EURO]: SHORT_CURRENCIES.EURO,
};
