// @celo/abis-12/dist/cjs/FederatedAttestations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.federatedAttestationsABI = void 0;
exports.federatedAttestationsABI = [
    {
        payable: false,
        stateMutability: 'nonpayable',
        type: 'constructor',
        inputs: [{ name: 'test', internalType: 'bool', type: 'bool' }],
    },
    {
        type: 'event',
        anonymous: false,
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32', indexed: true },
            { name: 'issuer', internalType: 'address', type: 'address', indexed: true },
            { name: 'account', internalType: 'address', type: 'address', indexed: true },
            { name: 'signer', internalType: 'address', type: 'address', indexed: false },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64', indexed: false },
            { name: 'publishedOn', internalType: 'uint64', type: 'uint64', indexed: false },
        ],
        name: 'AttestationRegistered',
    },
    {
        type: 'event',
        anonymous: false,
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32', indexed: true },
            { name: 'issuer', internalType: 'address', type: 'address', indexed: true },
            { name: 'account', internalType: 'address', type: 'address', indexed: true },
            { name: 'signer', internalType: 'address', type: 'address', indexed: false },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64', indexed: false },
            { name: 'publishedOn', internalType: 'uint64', type: 'uint64', indexed: false },
        ],
        name: 'AttestationRevoked',
    },
    {
        type: 'event',
        anonymous: false,
        inputs: [
            { name: 'eip712DomainSeparator', internalType: 'bytes32', type: 'bytes32', indexed: false },
        ],
        name: 'EIP712DomainSeparatorSet',
    },
    {
        type: 'event',
        anonymous: false,
        inputs: [
            { name: 'previousOwner', internalType: 'address', type: 'address', indexed: true },
            { name: 'newOwner', internalType: 'address', type: 'address', indexed: true },
        ],
        name: 'OwnershipTransferred',
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'EIP712_OWNERSHIP_ATTESTATION_TYPEHASH',
        outputs: [{ name: '', internalType: 'bytes32', type: 'bytes32' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'MAX_ATTESTATIONS_PER_IDENTIFIER',
        outputs: [{ name: '', internalType: 'uint256', type: 'uint256' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'MAX_IDENTIFIERS_PER_ADDRESS',
        outputs: [{ name: '', internalType: 'uint256', type: 'uint256' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [
            { name: '', internalType: 'address', type: 'address' },
            { name: '', internalType: 'address', type: 'address' },
            { name: '', internalType: 'uint256', type: 'uint256' },
        ],
        name: 'addressToIdentifiers',
        outputs: [{ name: '', internalType: 'bytes32', type: 'bytes32' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'eip712DomainSeparator',
        outputs: [{ name: '', internalType: 'bytes32', type: 'bytes32' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [
            { name: '', internalType: 'bytes32', type: 'bytes32' },
            { name: '', internalType: 'address', type: 'address' },
            { name: '', internalType: 'uint256', type: 'uint256' },
        ],
        name: 'identifierToAttestations',
        outputs: [
            { name: 'account', internalType: 'address', type: 'address' },
            { name: 'signer', internalType: 'address', type: 'address' },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64' },
            { name: 'publishedOn', internalType: 'uint64', type: 'uint64' },
        ],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'initialized',
        outputs: [{ name: '', internalType: 'bool', type: 'bool' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'isOwner',
        outputs: [{ name: '', internalType: 'bool', type: 'bool' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'owner',
        outputs: [{ name: '', internalType: 'address', type: 'address' }],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [],
        name: 'registryContract',
        outputs: [{ name: '', internalType: 'contract IRegistry', type: 'address' }],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [],
        name: 'renounceOwnership',
        outputs: [],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [{ name: '', internalType: 'bytes32', type: 'bytes32' }],
        name: 'revokedAttestations',
        outputs: [{ name: '', internalType: 'bool', type: 'bool' }],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [{ name: 'newOwner', internalType: 'address', type: 'address' }],
        name: 'transferOwnership',
        outputs: [],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [],
        name: 'initialize',
        outputs: [],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32' },
            { name: 'account', internalType: 'address', type: 'address' },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64' },
        ],
        name: 'registerAttestationAsIssuer',
        outputs: [],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32' },
            { name: 'issuer', internalType: 'address', type: 'address' },
            { name: 'account', internalType: 'address', type: 'address' },
            { name: 'signer', internalType: 'address', type: 'address' },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64' },
            { name: 'v', internalType: 'uint8', type: 'uint8' },
            { name: 'r', internalType: 'bytes32', type: 'bytes32' },
            { name: 's', internalType: 'bytes32', type: 'bytes32' },
        ],
        name: 'registerAttestation',
        outputs: [],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32' },
            { name: 'issuer', internalType: 'address', type: 'address' },
            { name: 'account', internalType: 'address', type: 'address' },
        ],
        name: 'revokeAttestation',
        outputs: [],
    },
    {
        constant: false,
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
        inputs: [
            { name: 'issuer', internalType: 'address', type: 'address' },
            { name: 'identifiers', internalType: 'bytes32[]', type: 'bytes32[]' },
            { name: 'accounts', internalType: 'address[]', type: 'address[]' },
        ],
        name: 'batchRevokeAttestations',
        outputs: [],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32' },
            { name: 'trustedIssuers', internalType: 'address[]', type: 'address[]' },
        ],
        name: 'lookupAttestations',
        outputs: [
            { name: 'countsPerIssuer', internalType: 'uint256[]', type: 'uint256[]' },
            { name: 'accounts', internalType: 'address[]', type: 'address[]' },
            { name: 'signers', internalType: 'address[]', type: 'address[]' },
            { name: 'issuedOns', internalType: 'uint64[]', type: 'uint64[]' },
            { name: 'publishedOns', internalType: 'uint64[]', type: 'uint64[]' },
        ],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [
            { name: 'account', internalType: 'address', type: 'address' },
            { name: 'trustedIssuers', internalType: 'address[]', type: 'address[]' },
        ],
        name: 'lookupIdentifiers',
        outputs: [
            { name: 'countsPerIssuer', internalType: 'uint256[]', type: 'uint256[]' },
            { name: 'identifiers', internalType: 'bytes32[]', type: 'bytes32[]' },
        ],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'pure',
        type: 'function',
        inputs: [],
        name: 'getVersionNumber',
        outputs: [
            { name: '', internalType: 'uint256', type: 'uint256' },
            { name: '', internalType: 'uint256', type: 'uint256' },
            { name: '', internalType: 'uint256', type: 'uint256' },
            { name: '', internalType: 'uint256', type: 'uint256' },
        ],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'view',
        type: 'function',
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32' },
            { name: 'issuer', internalType: 'address', type: 'address' },
            { name: 'account', internalType: 'address', type: 'address' },
            { name: 'signer', internalType: 'address', type: 'address' },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64' },
            { name: 'v', internalType: 'uint8', type: 'uint8' },
            { name: 'r', internalType: 'bytes32', type: 'bytes32' },
            { name: 's', internalType: 'bytes32', type: 'bytes32' },
        ],
        name: 'validateAttestationSig',
        outputs: [],
    },
    {
        constant: true,
        payable: false,
        stateMutability: 'pure',
        type: 'function',
        inputs: [
            { name: 'identifier', internalType: 'bytes32', type: 'bytes32' },
            { name: 'issuer', internalType: 'address', type: 'address' },
            { name: 'account', internalType: 'address', type: 'address' },
            { name: 'signer', internalType: 'address', type: 'address' },
            { name: 'issuedOn', internalType: 'uint64', type: 'uint64' },
        ],
        name: 'getUniqueAttestationHash',
        outputs: [{ name: '', internalType: 'bytes32', type: 'bytes32' }],
    },
];
